### PR TITLE
[#5112] feat(core): support pre event for event listener systems

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?
support pre event for event listener systems
1. add  new `PreEvent` to represent Pre event and only `SYNC` event listeners could process `PreEvent`
2. keep `Event` as post event to keep compatibility.
3. `EventBus` dispatch event to corresponding event listeners.

### Why are the changes needed?
Fix: #5112 

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
add UT

### CODE REVIEW

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java [hunk 1]
```diff
@@ -95,17 +97,29 @@ enum Mode {
...
    */
-  void onPostEvent(Event event) throws RuntimeException;
+  default void onPostEvent(Event postEvent) throws RuntimeException {}
+
+  /**
+   * Handles pre events generated before the operation.
+   *
+   * <p>This method provides a hook for pre-operation event processing, any changes you make to the
+   * resources during the event will be reflected in the operations that follow. You must set mode
+   * to SYNC to process pre events.
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java:116]: Why do we need to set SYNC to support pre events?

File: core/src/main/java/org/apache/gravitino/listener/EventBus.java [hunk 2]
```diff
@@ -35,25 +37,45 @@ public class EventBus {
   // which are meant for synchronous event listening, or AsyncQueueListener, designed for
   // asynchronous event processing.
   private final List<EventListenerPlugin> postEventListeners;
+  // Only contain sync EventListenerPlugins.
+  private final List<EventListenerPlugin> preEventListeners;
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/EventBus.java:41]: Why do we need to have 2 listeners for pre and post event separately?

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java (hunk 1 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java]: removed

File: core/src/main/java/org/apache/gravitino/listener/EventBus.java (hunk 2 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/EventBus.java]: removed

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java [hunk 3]
```diff
@@ -95,17 +97,29 @@ enum Mode {
...
+   * resource in the event.
    *
-   * @param event The event to be processed.
-   * @throws RuntimeException Indicates issues encountered during event processing.
+   * @param postEvent The post event to be processed.
+   * @throws RuntimeException Indicates issues encountered during event processing, this has no
+   *     affect to the operation.
    */
-  void onPostEvent(Event event) throws RuntimeException;
+  default void onPostEvent(Event postEvent) throws RuntimeException {}
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java:109]: Why do we change to `default`?

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java [hunk 4]
```diff
@@ -95,17 +97,29 @@ enum Mode {
   void stop() throws RuntimeException;
 
   /**
-   * Handles events generated after the completion of an operation. Implementers are responsible for
-   * processing these events, which may involve additional logic to respond to the operation
-   * outcomes.
+   * Handles events generated after the completion of an operation.
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java:100]: "Handle post-events..."

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java [hunk 5]
```diff
@@ -95,17 +97,29 @@ enum Mode {
...
-   * @throws RuntimeException Indicates issues encountered during event processing.
+   * @param postEvent The post event to be processed.
+   * @throws RuntimeException Indicates issues encountered during event processing, this has no
+   *     affect to the operation.
    */
-  void onPostEvent(Event event) throws RuntimeException;
+  default void onPostEvent(Event postEvent) throws RuntimeException {}
+
+  /**
+   * Handles pre events generated before the operation.
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java:112]: "Handle pre-events..."

File: core/src/main/java/org/apache/gravitino/listener/EventBus.java [hunk 6]
```diff
@@ -64,7 +73,15 @@ public void dispatchEvent(Event event) {
...
+  List<EventListenerPlugin> getEventListeners() {
+    return eventListeners;
+  }
+
+  private void dispatchPostEvent(Event postEvent) {
+    eventListeners.forEach(postEventListener -> postEventListener.onPostEvent(postEvent));
+  }
+
+  private void dispatchPreEvent(PreEvent preEvent) throws ForbiddenException {
+    eventListeners.forEach(preEventListener -> preEventListener.onPreEvent(preEvent));
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/EventBus.java:85]: Actually, we don't differentiate pre or post event listeners, the code here is a little confusing. We can change to:

```java
  private void dispatchPostEvent(Event postEvent) {
    eventListeners.forEach(listener -> listener.onPostEvent(postEvent));
  }
```

File: core/src/main/java/org/apache/gravitino/listener/EventListenerPluginWrapper.java [hunk 7]
```diff
@@ -68,13 +70,40 @@ public void onPostEvent(Event event) {
...
+  public void onPreEvent(PreEvent preEvent) {
+    try {
+      userEventListener.onPreEvent(preEvent);
+    } catch (ForbiddenException e) {
+      if (Mode.SYNC.equals(mode())) {
+        LOG.warn(
+            "Event listener {} process pre event {} failed, will skip the operation.",
+            listenerName,
+            preEvent.getClass().getSimpleName(),
+            e);
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/EventListenerPluginWrapper.java:90]: I think you can extract the common part here to simplify the code.

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java [hunk 8]
```diff
@@ -95,17 +97,29 @@ enum Mode {
...
+   *
+   * <p>This method handles pre-operation events in SYNC or ASYNC mode, any changes to resources in
+   * the event will inflect the subsequent operations.
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java:115]: "will affect..."

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java [hunk 9]
```diff
@@ -95,17 +97,29 @@ enum Mode {
...
+   *
+   * @param preEvent The pre event to be processed.
+   * @throws ForbiddenException The operation will be skipped if and only if throwing {@code
+   *     org.apache.gravitino.exceptions.ForbiddenException} and the event listener is SYNC mode,
+   *     the exception will be ignored in other conditions.
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java:120]: Can you please explain about the exception here, how you handle this?

File: core/src/main/java/org/apache/gravitino/listener/EventBus.java (hunk 6 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/EventBus.java:85]: ok

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java (hunk 3 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java:109]: after supporting pre event,  the user who only interest on pre event, could only implement onPreEvent interface, no need to implement onPostEvent interface

File: core/src/main/java/org/apache/gravitino/listener/EventListenerPluginWrapper.java (hunk 7 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/EventListenerPluginWrapper.java]: done

File: core/src/main/java/org/apache/gravitino/listener/EventBus.java (hunk 6 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/EventBus.java]: done

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java (hunk 5 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java]: updated

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java (hunk 4 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java]: updated

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java (hunk 8 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java]: done

File: core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java (hunk 9 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/api/EventListenerPlugin.java]: updated

**Review by jerryshao — APPROVED** [review]

**Pull request merged**